From 2b2f8c4e374f2edc535abc0a48edb2b3f02dd538 Mon Sep 17 00:00:00 2001
From: Raphael Gallais-Pou <raphael.gallais-pou@foss.st.com>
Date: Mon, 25 Sep 2023 10:39:35 +0200
Subject: [PATCH 0586/1141] drm/stm: dsi: handle loopback pixel clock

Loopback pixel clock needs to be handled, when enabling the DSI. Since
the common clock framework has no idea of setting parent when a
connector is enabled, it must be done manually in the connector hook.

Signed-off-by: Raphael Gallais-Pou <raphael.gallais-pou@foss.st.com>
Change-Id: I5ffdeb13fd0a6b542443887552be77a7e62b50ce
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/330817
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Yannick FERTRE <yannick.fertre@foss.st.com>
Domain-Review: Yannick FERTRE <yannick.fertre@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
---
 drivers/gpu/drm/stm/dw_mipi_dsi-stm.c | 21 +++++++++++++++++++++
 1 file changed, 21 insertions(+)

--- a/drivers/gpu/drm/stm/dw_mipi_dsi-stm.c
+++ b/drivers/gpu/drm/stm/dw_mipi_dsi-stm.c
@@ -134,6 +134,8 @@ struct dw_mipi_dsi_stm {
 	struct device *dev;
 	struct clk *pllref_clk;
 	struct clk *pclk;
+	struct clk *px_clk;
+	struct clk *ltdc_clk;
 	struct clk_hw txbyte_clk;
 	struct dw_mipi_dsi *dsi;
 	u32 hw_version;
@@ -386,6 +388,13 @@ static int dw_mipi_dsi_phy_141_init(void
 
 	DRM_DEBUG_DRIVER("\n");
 
+	/* Switch pixel clock parent to ref clock */
+	ret = clk_set_parent(dsi->ltdc_clk, dsi->px_clk);
+	if (ret) {
+		DRM_ERROR("Could not set pixel clock parent: %d\n", ret);
+		return ret;
+	}
+
 	/* Select video mode by resetting DSIM bit */
 	dsi_clear(dsi, DSI_WCFGR, WCFGR_DSIM);
 
@@ -1182,6 +1191,18 @@ static int dw_mipi_dsi_stm_probe(struct
 		goto err_dsi_probe;
 	}
 
+	dsi->px_clk = devm_clk_get(dev, "px_clk");
+	/* No need to return since only MP25 has it */
+	if (IS_ERR(dsi->px_clk))
+		dev_err_probe(dev, PTR_ERR(dsi->px_clk), "Unable to get px_clk clock\n");
+
+	if (dsi->hw_version == HWVER_141) {
+		dsi->ltdc_clk = devm_clk_get(dev, "pixclk");
+		/* No need to return since only MP25 has it */
+		if (IS_ERR(dsi->ltdc_clk))
+			dev_err_probe(dev, PTR_ERR(dsi->ltdc_clk), "Unable to get pixclk clock\n");
+	}
+
 	clk_disable_unprepare(dsi->pclk);
 
 	return 0;
